### PR TITLE
NIFI-461 Add processors to interact with Apache Solr

### PR DESCRIPTION
**Pull request by bbende**
This pull request provides two processors:
* GetSolr - extracts data from Solr by running a query and sending the results out in XML, the intent is to run on a timer, or cron, and only get the results since last execution, similar to GetHttp.

* PutSolrContentStream - streams the content of a FlowFile to a Solr update handler, the main uses cases would be indexing json, xml, and csv, but any update handler could be used. MergeContent could be used in front of this processor to combine documents for more efficient submission to Solr. A good overview of how Solr indexes JSON is here: http://lucidworks.com/blog/indexing-custom-json-data/ and this processor could be used to post the JSON.

Both processors support a stand-alone Solr instance, as well as a SolrCloud cluster.

### CODE REVIEW

**Comment by joewitt** [comment]
@madrob Can you take a look?

File: nifi/nifi-nar-bundles/nifi-solr-bundle/pom.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        <version>0.1.0-incubating-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>nifi-solr-bundle</artifactId>
+    <packaging>pom</packaging>
+
+    <description>A bundle of processors that can store and retrieve data from Apache Solr</description>
+
+    <properties>
+        <solr.version>5.0.0</solr.version>
```
Comment (madrob) [inline nifi/nifi-nar-bundles/nifi-solr-bundle/pom.xml]: 5.1.0 was released this week, no reason not to use that instead.

File: nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+     *
+     * @param context
+     *          The context
+     * @return an HttpSolrClient or CloudSolrClient
+     */
+    protected SolrClient createSolrClient(final ProcessContext context) {
+        if (SOLR_TYPE_STANDARD.equals(context.getProperty(SOLR_TYPE).getValue())) {
+            return new HttpSolrClient(context.getProperty(SOLR_LOCATION).getValue());
+        } else {
+            CloudSolrClient cloudSolrServer = new CloudSolrClient(
```
Comment (madrob) [inline nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: Nit: This naming is potentially confusing. Variable name could be cloudSolrClient.

File: nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                    context.getProperty(COLLECTION).evaluateAttributeExpressions().getValue());
+            return cloudSolrServer;
+        }
+    }
+
+    /**
+     * Returns the {@link org.apache.solr.client.solrj.SolrClient} that was created by the
+     * {@link #createSolrClient(org.apache.nifi.processor.ProcessContext)} method
+     *
+     * @return
```
Comment (madrob) [inline nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: Nit: empty javadoc.

File: nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/pom.xml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.solr</groupId>
+            <artifactId>solr-core</artifactId>
+            <version>${solr.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <!-- These Lucene deps should be brought in through solr-core, but there
+         appears to be an issue with 5.0.0 that still references some 4.10.3 poms -->
```
Comment (madrob) [inline nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/pom.xml]: Is this still an issue with 5.1?

**Comment by madrob** [comment]
First a caveat - I'm less familiar with the mechanics of the ContentStreamUpdateRequest, but it looks like this should work fine. Does this handle all of insert/update/delete, or just insert? I'm more used to seeing separate docs added like you do in the test code for the GetSolr, which is what I expected the PutSolr to look like.

+1 from me.

File: nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/test/java/org/apache/nifi/processors/solr/TestGetSolr.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            doc4.addField("last", "smith");
+            doc4.addField("created", new Date());
+
+            SolrInputDocument doc5 = new SolrInputDocument();
+            doc5.addField("first", "joan");
+            doc5.addField("last", "smith");
+            doc5.addField("created", new Date());
+
+            // add the test data to the index
+            solrClient.add(doc1);
```
Comment (madrob) [inline nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/test/java/org/apache/nifi/processors/solr/TestGetSolr.java]: Can use a collection here.

File: nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/pom.xml (hunk 4 above)
Comment (bbende) [inline nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/pom.xml]: It looks like it still is, here are the dependencies brought in through solr-core 5.1.0:

+- org.apache.solr:solr-core:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-analyzers-common:jar:4.10.3:test
[INFO] |  +- org.apache.lucene:lucene-analyzers-kuromoji:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-analyzers-phonetic:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-backward-codecs:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-codecs:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-core:jar:4.10.3:test
[INFO] |  +- org.apache.lucene:lucene-expressions:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-grouping:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-highlighter:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-join:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-memory:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-misc:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-queries:jar:5.1.0:test
[INFO] |  +- org.apache.lucene:lucene-queryparser:jar:4.10.3:test

I looked at the solr-core 5.1.0 pom, but I don't see why that is happening, and it looks like it got pulled from Maven central:

#Thu Apr 16 20:46:02 EDT 2015
solr-core-5.1.0.jar>central=
solr-core-5.1.0.pom>central=

Comment (madrob) [inline nifi/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/pom.xml]: I would suggest filing an issue at https://issues.apache.org/jira/browse/SOLR and see what response shakes out.

**Comment by bbende** [comment]
Thanks for reviewing Mike. 

With regards to the ContentStreamUpdateRequest... I actually hadn't thought about using it for deletes, but I just did a test and it worked. So it should work for almost any add/update/delete example in the Solr documentation. It would just require an upstream processor that was producing the right commands, where as for the case of adding new documents it seems more likely that you already have some data (json, csv, etc) that you can just fork into Solr.

I was thinking there eventually might be a use case for another type of "PutSolr" processor that reads in data in a standard format (maybe avro) and converted to a SolrInputDocument, but seemed like the ContentStream approach gave the most flexibility initially.

**Comment by madrob** [comment]
> I actually hadn't thought about using it for deletes, but I just did a test and it worked.
Awesome! Can you add that test to the Unit Tests?

> I was thinking there eventually might be a use case for another type of "PutSolr" processor that reads in data in a standard format (maybe avro) and converted to a SolrInputDocument, but seemed like the ContentStream approach gave the most flexibility initially.
Agreed. Just file a follow-on issue and somebody (maybe you) will get to it when the need arises.

**Comment by bbende** [comment]
Added a test case that shows a delete by query using xml, ex:
<delete>
  <query>...</query>
</delete>

**Comment by bbende** [comment]
<delete>
      <query>...</query>
    </delete>

**Pull request merged**